ci/travis/ci.sh

```diff
@@ -38,21 +38,187 @@ should_run_job() {
   set +x
-  local to_add="$HOME/miniconda/bin" old_path=":${PATH}:"
-  if [ "${old_path}" = "${old_path##*:${to_add}:*}" ]; then
-    PATH="${to_add}:$PATH"
+  local old="${PATH}"
+  if [ "${OSTYPE}" = msys ]; then
+    PATH="${PATH// :/:}"  # HACK: Work around https://github.com/actions/virtual-environments/issues/635#issue-589318442
+  fi
+
+  local to_add
+  for to_add in "$@"; do
+    case ":${PATH}:" in
+      *:"${to_add}":*) ;;  # Do nothing; already in PATH
+      *) PATH="${to_add}:${PATH}" ;;  # Prepend to PATH
+    esac
+  done
+
+  if [ "${old}" != "${PATH}" ]; then
+    echo "PATH=${PATH}"
   fi
   test -z "${set_x}" || set -x  # restore set -x
+}
+
+test_python() {
   if [ "${OSTYPE}" = msys ]; then
-    PATH="${PATH// :/:}"  # HACK: Work around https://github.com/actions/virtual-environments/issues/635#issue-589318442
+    # Windows -- most tests won't work yet; just do the ones we know work
+    PYTHONPATH=python python -m pytest --durations=5 --timeout=300 python/ray/tests/test_mini.py
+  fi
+}
+
+build_sphinx_docs() {
+  (
+    cd "${WORKSPACE_DIR}"/doc
+    if [ "${OSTYPE}" = msys ]; then
+      echo "WARNING: Documentation not built on Windows due to currently-unresolved issues"
+    else
+      sphinx-build -q -W -E -T -b html source _build/html
+    fi
+  )
+}
+
+lint_readme() {
+  (
+    cd "${WORKSPACE_DIR}"/python
+    python setup.py check --restructuredtext --strict --metadata
+  )
+}
+
+lint_python() {
+  # ignore dict vs {} (C408), others are defaults
+  command -V python
+  python -m flake8 --inline-quotes '"' --no-avoid-escape --exclude=python/ray/core/generated/,streaming/python/generated,doc/source/conf.py,python/ray/cloudpickle/,python/ray/thirdparty_files --ignore=C408,E121,E123,E126,E226,E24,E704,W503,W504,W605
+  "${ROOT_DIR}"/format.sh --all
+}
+
+lint_bazel() {
+  local platform="" architecture="" suffix=".tar.gz"
+  case "${OSTYPE}" in
+    linux*) platform=linux;;
+    darwin*) platform=darwin;;
+    msys*) platform=windows; suffix=".zip";;
+  esac
+  case "${HOSTTYPE}" in
+    x86_64) architecture=amd64;;
+    i?86*) architecture=i386;;
+  esac
+  # Run buildifier without affecting external environment variables
+  (
+    # TODO: Move installing Go & building buildifier to the dependency installation step?
+    if [ ! -d "${GOROOT}" ]; then
+      local url="https://dl.google.com/go/go1.14.2.${platform}-${architecture}${suffix}"
+      local filename="${GOROOT%/*}/${url##*/}"
+      if [ "${suffix}" = ".zip" ]; then
+        curl -s -L "${url}" -o "${filename}"
+        unzip -q -d "${GOROOT%/*}" -- "${filename}"
+        rm -f -- "${filename}"
+      else
+        curl -s -L "${url}" | tar -C "${GOROOT%/*}" -xz
+      fi
+    fi
+    mkdir -p -- "${GOPATH}"
+    prepend_PATH "${GOPATH}/bin" "${GOROOT}/bin"
+
+    # Build buildifier
+    go get github.com/bazelbuild/buildtools/buildifier
+
+    # Now run buildifier
+    "${ROOT_DIR}"/bazel-format.sh
+  )
+}
+
+lint_web() {
```

Comment:
let's keep the linting code we had before here (especially, get rid of all the if things here)